gobblin-data-management/src/main/java/org/apache/gobblin/data/management/conversion/hive/source/HiveAvroToOrcSource.java

```diff
@@ -36,8 +43,36 @@
     if (!state.contains(HiveDatasetFinder.HIVE_DATASET_CONFIG_PREFIX_KEY)) {
       state.setProp(HiveDatasetFinder.HIVE_DATASET_CONFIG_PREFIX_KEY, "hive.conversion.avro");
     }
+    this.lineageInfo = LineageInfo.getLineageInfo(state.getBroker());
+
+    List<WorkUnit> workunits = super.getWorkunits(state);
+    for (WorkUnit workUnit : workunits) {
+      if (shouldSetLineageInfo(workUnit)) {
+        setSourceLineageInfo(workUnit, (ConvertibleHiveDataset) ((HiveWorkUnit) workUnit).getHiveDataset(),
+            this.lineageInfo);
+      }
+    }
+    return workunits;
+  }
 
-    return super.getWorkunits(state);
+  @VisibleForTesting
+  public static void setSourceLineageInfo(WorkUnit workUnit, ConvertibleHiveDataset hiveDataset,
+      Optional<LineageInfo> lineageInfo) {
+    DatasetDescriptor sourceDataset = hiveDataset.getSourceDataset();
+    if (lineageInfo.isPresent()) {
+      lineageInfo.get().setSource(sourceDataset, workUnit);
+    }
   }
 
-}
+  public static boolean shouldSetLineageInfo(WorkUnit workUnit) {
```

Comment:
How about `public static ConvertibleHiveDataset getConvertibleHiveDataset(WorkUnit workUnit)`, which returns the casted `convertibleHiveDataset` and `null` if not possible. Then we can have a cleaner logic
```
setSourceLineageInfo(workUnit, getConvertibleHiveDataset(workUnit), this.lineageInfo);
```
Remember to check null for `hiveDataset` in `setSourceLineageinfo`

Reply:
I actually like the shouldSetLineageInfo function - although I think, like your other comments, it should not be a static method just for unit test purposes